packages/grafana-ui/src/components/uPlot/config/UPlotSeriesBuilder.ts

```diff
@@ -1,31 +1,267 @@
 import tinycolor from 'tinycolor2';
-import { Series } from 'uplot';
-import { PointMode } from '../config';
+import uPlot, { Series } from 'uplot';
+import { GraphMode, LineConfig, AreaConfig, PointsConfig, PointMode, LineInterpolation } from '../config';
 import { PlotConfigBuilder } from '../types';
 
-export interface SeriesProps {
+export interface SeriesProps extends LineConfig, AreaConfig, PointsConfig {
+  mode: GraphMode;
   scaleKey: string;
-  line?: boolean;
-  lineColor?: string;
-  lineWidth?: number;
-  points?: PointMode;
-  pointSize?: number;
-  pointColor?: string;
-  fill?: boolean;
-  fillOpacity?: number;
-  fillColor?: string;
+}
+
+function buildBarsPaths(u: uPlot, seriesIdx: number, idx0: number, idx1: number): Series.Paths {
+  let series = u.series[seriesIdx];
+  let scaleKey = series.scale as string;
+
+  const gapPct = 0.1;
+
+  let gap = Math.round((u.width * gapPct) / (idx1 - idx0));
+  let maxWidth = Infinity;
+
+  //@ts-ignore
+  let fillTo = series.fillTo(u, seriesIdx, series.min, series.max);
+
+  let y0Pos = Math.round(u.valToPos(fillTo, scaleKey, true));
+  let colWid = u.bbox.width / (idx1 - idx0);
+
+  let stroke = new Path2D();
+
+  for (let i = idx0; i <= idx1; i++) {
+    let yVal = u.data[seriesIdx][i];
+
+    if (yVal == null) {
+      continue;
+    }
+
+    let xVal = u.scales.x.distr === 2 ? i : u.data[0][i];
+
+    // TODO: all xPos can be pre-computed once for all series in aligned set
+    let xPos = u.valToPos(xVal, 'x', true);
+    let yPos = u.valToPos(yVal, scaleKey, true);
+
+    let wid = Math.min(maxWidth, colWid - gap);
+    let lft = xPos - wid / 2;
+    let btm = Math.min(yPos, y0Pos);
+    let top = Math.max(yPos, y0Pos);
+
+    stroke.rect(Math.round(lft), Math.round(top), Math.round(wid), Math.round(btm - top));
+  }
+
+  let fill = series.fill != null ? new Path2D(stroke) : undefined;
+
+  return {
+    stroke,
+    fill,
+  };
+}
+
+function buildStaircasePaths(u: uPlot, seriesIdx: number, idx0: number, idx1: number): Series.Paths {
```

Comment:
would move this and the catmullRomFitting func to a lineInterpolation file

Reply:
moved out to `paths.ts`